### ISSUE TITLE
JavadocType: False negative for unused `@param` tag for Type Parameters enclosing illegal whitespace

### ISSUE THREAD

**Issue by sktpy** [checkstyle/checkstyle#14581]
**Test.java**
```java
/**
 *
 * @param < T > first type
 * @param <U> second type
 */
public class Test<T, U> {}
```
<br>

**config.xml**
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
</module>
```
<br>

**Current CLI:**
```
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:6:1: Type Javadoc comment is missing @param <T> tag. [JavadocType]
Audit done.
Checkstyle ends with 1 errors.
```
<br>

**Expected:**
```
Starting audit...
[ERROR] Test.java:3:4: Unused @param tag for '<'. [JavadocType]
[ERROR] Test.java:6:1: Type Javadoc comment is missing @param <T> tag. [JavadocType]
Audit done.
Checkstyle ends with 2 errors.
```

Note: This expected error is received if the `"first type"` param name in doc does not matches with type name of class.
<br>

**Explanation (a bit off topic)** 
- This bug came in the crossfire when I was implementing a cleaner logic for extracting param name from tag to resolve https://github.com/checkstyle/checkstyle/issues/13309.
- But, I am unable to proceed in order to preserve this false negative, which can be found in this input file:
https://github.com/checkstyle/checkstyle/blob/83c88593c82d5d59f803474d134f8c50394daa2f/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadoctype/InputJavadocTypeTestTrimProperty.java#L19
- If we do not preserve this false negative, we are able to get much cleaner logic due to the invariant it grants i.e. param name cannot contain whitespace.
- Param name really cannot contain whitespace, this code produces error in javadoc generation `Test.java:3: error: '>' expected`
- Which is also confirmed in the rendered doc, as `T - first type` does not gets included:
![image](https://github.com/checkstyle/checkstyle/assets/145206423/974cbc53-3397-45ef-a5ce-60f908ab7a3d)

**Comment by sktpy** [comment]
I want to fix it in the referenced issue, in order to have cleaner logic for extracting param name from tag.

I have also implemented the fix where we are able to preserve this false negative but I haven't created a PR as it wouldn't make much sense to do that and then say I don't want it this way :joy:

**Comment by nrmancuso** [comment]
I have been thinking about this one, and honestly I am having a hard time approving this without soome more information. @sktpy please make sure I understand this correctly:

 - `@param < T >` is not a valid param tag, and throws an error in javadoc generation
 - this is due to the javadoc tool thinking that `<` is the identifier here, which is not a valid identifier (by  java's own definition)
 - javadoc generation still happens (doesn't fail) but this tag is completely ignored

I think we need to take a step back, and look at the AST generated for this javadoc first. Let's see how checkstyle parses this compared to a valid tag.

If we do indeed parse this like `<` is the identifier, it might be better for us to throw an error like the javadoc tool would to avoid processing this at all.

Let's consider https://checkstyle.org/writingchecks.html#Limitations:

>Code has to be compilable by javac to get valid violations. If it is not, you can get hard to understand parse errors.

Perhaps we need to extend this list to include invalid javadocs.

Action item: @sktpy please generate ASTs for valid and invalid tags like in your example, and provide an analysis here.

**Comment by sktpy** [comment]
**Example Used:**
```java
/**
 * @param < T > first
 * @param <> second
 * @param <*> third
 * @param <S> fourth
 * @param <U> fifth
 * @param < V > sixth
 */
public class Test<T, U> {}
```
<hr><br>

> @param < T > is not a valid param tag, and throws an error in javadoc generation

Yes.
<hr><br>

> this is due to the javadoc tool thinking that < is the identifier here, which is not a valid identifier (by java's own definition)

Kind of yes, but more precisely judging by the error messages below:
```java
Test.java:2: error: '>' expected
 * @param < T > first
   ^
Test.java:3: error: identifier expected
 * @param <> second
   ^
Test.java:4: error: identifier expected
 * @param <*> third
   ^
Test.java:5: error: @param name not found
 * @param <S> fourth
           ^
Test.java:7: error: '>' expected
 * @param < V > sixth
   ^
```
(I haven't looked at the javadoc source code yet, so the following is a guess)
It seems like when opening angle bracket is encountered,
javadoc first makes sure that it is ended with closing angle bracket without encountering any whitespace,
after that it is checked for valid identifier, and then if the valid identifier is actually present.
<hr><br>

> javadoc generation still happens (doesn't fail) but this tag is completely ignored

Yes expect for when there is error for not-found/unused valid identifier, rest of the tag with errors above are ignored.
![image](https://github.com/checkstyle/checkstyle/assets/145206423/fd1d6815-fd01-4056-a34b-c1c17e39f1bc)
<hr><br>

> we need to take a step back, and look at the AST

AST generated by Checkstyle:
```java
  ... 
      ...
       |--BLOCK_COMMENT_BEGIN -> /* [1:0]
       |   |--COMMENT_CONTENT -> *\n * @param < T > first\n * @param <> second\n * @param <*> third\n * @param <S> fourth\n * @param <U> fifth\n  [1:2]
       |   `--BLOCK_COMMENT_END -> */ [7:1]
       ...
   ...
```
There isn't much to look here, valid or invalid, all the tags are generated in the same way, the only difference being the literal content.
Actual processing on how these are handled are done by the check and utils used.
<hr><br>

> Code has to be compilable by javac to get valid violations.

Another thing regarding this is that currently violation is produced exactly as specified **Expected** in the issue description,
but due to the buggy implementation, this only happens specifically when the invalid-param name, is not present in the def:
```java
java -jar checkstyle-10.14.0-all.jar -c test/config.xml test/Test.java
Starting audit...
[ERROR] /home/skt/Documents/oss/sandbox/test/Test.java:3:4: Unused @param tag for '<>'. [JavadocType]
[ERROR] /home/skt/Documents/oss/sandbox/test/Test.java:4:4: Unused @param tag for '<*>'. [JavadocType]
[ERROR] /home/skt/Documents/oss/sandbox/test/Test.java:5:4: Unused @param tag for '<S>'. [JavadocType]
[ERROR] /home/skt/Documents/oss/sandbox/test/Test.java:7:4: Unused @param tag for '<'. [JavadocType] // violation in question
[ERROR] /home/skt/Documents/oss/sandbox/test/Test.java:9:1: Type Javadoc comment is missing @param <T> tag. [JavadocType]
Audit done.
Checkstyle ends with 5 errors.
```
As we can see, although the sixth and the first param are exactly the same, but we only got violation for the sixth one, because it was not present in class's parameters. (will leave a comment regarding why exactly this happens in the PR of the alternative fix that I have created in my fork _shortly_)

**Comment by nrmancuso** [comment]
> AST generated by Checkstyle

Make sure to use `-J` to see Javadoc AST.

**Comment by sktpy** [comment]
Apologies, had recalled `-T` from the memory which was actually for comments not javadoc, should have referred to the doc ;)

**Invalid:** `@param < T > first`
```cs
    |--JAVADOC_TAG -> JAVADOC_TAG [2:3]
    |   |--PARAM_LITERAL -> @param [2:3]
    |   |--WS ->   [2:9]
    |   |--PARAMETER_NAME -> < [2:10]
    |   |--WS ->   [2:11]
    |   `--DESCRIPTION -> DESCRIPTION [2:12]
    |       |--TEXT -> T > first [2:12]
    |       `--NEWLINE -> \n [2:21]
 ```
<hr>

**Valid:** `@param <U> fifth`
```cs
    |--JAVADOC_TAG -> JAVADOC_TAG [5:3]
    |   |--PARAM_LITERAL -> @param [5:3]
    |   |--WS ->   [5:9]
    |   |--PARAMETER_NAME -> <U> [5:10]
    |   |--WS ->   [5:13]
    |   `--DESCRIPTION -> DESCRIPTION [5:14]
    |       |--TEXT -> fifth [5:14]
    |       `--NEWLINE -> \n [5:19]
```
<hr>

- In the AST, the first whitespace after opening angle bracket is considered the splitting point, left part going to the `PARAMETER_NAME` and the right part going to the `TEXT`.

- For **invalid** `@param` usage (`first`), only the opening angle bracket is considered as the `PARAMETER_NAME`, everything rest goes to the `TEXT`.

- For **valid** usage (`second`), having no illegal whitespace, whole `<U>` is considered `PARAMETER_NAME`, and rest being the description text, as intended.

- `second`, `fourth`, `sixth` are similar with the difference being the value for their `PARAMETER_NAME` being `<>`, `<S>`, `<` respectively, remainder constitutes the `TEXT`.
<hr>

Interesting thing happens with the **third** (`@param <*> third`) in the input file:
```
Exception in thread "main" java.lang.IllegalArgumentException: [ERROR:4] Javadoc comment at column 11 has parse error. Details: mismatched input '*' expecting <EOF> while parsing JAVADOC
        at com.puppycrawl.tools.checkstyle.DetailNodeTreeStringPrinter.parseJavadocAsDetailNode(DetailNodeTreeStringPrinter.java:70)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.parseAndPrintJavadocTree(AstTreeStringPrinter.java:116)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:97)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:79)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:312)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
```
Caused by the asterisk being used as the "identifier" in the example, had to remove this input momentarily to generate AST for the rest.

**Comment by nrmancuso** [comment]
Ok @sktpy, I missed a point here. JavadocTypeCheck does not use the Javadoc AST; it parses comment content via regexp. So - we don't have to worry about any parsing or validation in that respect.

I think we should not pursue this issue; since it is not a defect in our parser, and such a javadoc is not valid, we should continue to just ignore such tags. Also consider:

>[ERROR] Test.java:3:4: Unused @param tag for '<'. [JavadocType]

This error would essentially be telling our users to create some parameter named `<`, which would not compile.

**Comment by sktpy** [comment]
> This error would essentially be telling our users to create some parameter named <, which would not compile

But we already have such error currently, just not for this specific input as explained in https://github.com/sktpy/checkstyle/pull/1#discussion_r1511710537

Consider the following input:
**Test.java**
```java
/**
 * @param < T > Hi, I am in class type!
 * @param < U > Hi, I am not in class type!
 */
public class Test<T> {}
```

**config.xml:**
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType">
      <property name="allowMissingParamTags" value="true"/>
    </module>
  </module>
</module>
```

**Current CLI:**
```
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:3:4: Unused @param tag for '<'. [JavadocType]
Audit done.
Checkstyle ends with 1 errors.
```

Although I don't believe this was the intended behaviour, but we should have violation for all such cases,
or don't have violation for such cases at all.

**Comment by sktpy** [comment]
Since, we don't check for violations on the input that aren't compilable, in my opinion we shouldn't have any violation for all such cases.
